File: Week 4/rust/7.interesting_alphabet.rs
/*
 * WEEK 4 - RUST PATTERN PROBLEMS
 * Topic: Interesting Alphabet Pattern
 * File: 7.interesting_alphabet.rs
 *
 * PATTERN (N=4): D / CD / BCD / ABCD
 */

use std::io::Read;

fn main() {
    let mut buf = String::new();
    std::io::stdin().read_to_string(&mut buf).ok();
    let n: usize = buf.split_whitespace().next().and_then(|s| s.parse().ok()).unwrap_or(4);

    for i in 1..=n {
        let start = b'A' + (n - i) as u8;
        for j in 0..i {
            print!("{}", (start + j as u8) as char);
        }
        println!();
    }
}

/*
 * NOTES:
 *  - Start letter shifts EARLIER as i grows; bottom row begins at 'A'.
 *  - Same column-offset pattern as char_pattern; only the start formula differs.
 */

File: Week 4/cpp/7.interesting_alphabet.cpp
/*
 * WEEK 4 - C++ PATTERN PROBLEMS
 * Topic: Interesting Alphabet Pattern
 * File: 7.interesting_alphabet.cpp
 *
 * PATTERN (N=4):
 *  D
 *  CD
 *  BCD
 *  ABCD
 */

#include <iostream>

int main() {
    int n;
    if (!(std::cin >> n)) { n = 4; std::cout << "(demo n=4)\n"; }
    for (int i = 1; i <= n; ++i) {
        char start = static_cast<char>('A' + n - i);
        for (int j = 0; j < i; ++j) {
            std::cout << static_cast<char>(start + j);
        }
        std::cout << '\n';
    }
    return 0;
}

/*
 * NOTES:
 *  - Start letter shifts EARLIER each row; bottom row begins at 'A'.
 *  - Same column-offset pattern as char_pattern, with a different start formula.
 */
